Don't generate checkout conflicts for submodules.

hackhaslam opened the pull request:
Checkout doesn't ever actually modify submodules in the workdir (does it?), so it shouldn't generate conflicts. Currently checkout generates a conflict if the submodule is dirty or if the submodule commit doesn't match the baseline. Neither of these match the behavior of git.git.

This change removes that logic and assumes that submodules aren't modified by checkout.

srajko commented:
I used a similar fix (definitely a hack in my case) to get around the problem - limited to an oid mismatch only: https://github.com/srajko/libgit2/commit/e07639fb8c7936feb864efd54ac619ea9790ff22

Going beyond that and ignoring wd changes (like in this PR) is probably the right thing to do for the case when the submodule is there both before and after the checkout.  I'm not sure about other cases - mostly worried about the `GIT_DELTA_TYPECHANGE` case here - https://github.com/libgit2/libgit2/blob/e8feafe32007ebd16a61820c70abd221655d053c/src/checkout.c#L491 and `GIT_DELTA_DELETED` here - https://github.com/libgit2/libgit2/blob/e8feafe32007ebd16a61820c70abd221655d053c/src/checkout.c#L462

Also, will `GIT_CHECKOUT_UPDATE_SUBMODULES` and `GIT_CHECKOUT_UPDATE_SUBMODULES_IF_CHANGED` affect this in the future?

hackhaslam commented:
Yeah, I need to investigate what git does in those cases. I guess that there are actually a bunch of permutations to consider between baseline, target and working directory.

I saw `GIT_CHECKOUT_UPDATE_SUBMODULES` and `GIT_CHECKOUT_UPDATE_SUBMODULES_IF_CHANGED` and I wondered if the conflict logic was looking forward to those options getting added. If that's the case then maybe the current logic should just be hidden behind those flags instead of removed.

hackhaslam commented:
I added some notes to the checkout internals doc. The new cases describe how git treats submodules during checkout. I haven't yet tried to determine where the behavior of libgit2 differs. I'll do that next.

srajko commented:
@hackhaslam thanks for pushing this forward.  If there is anything I can help with to keep this going please let me know.

srajko commented:
@hackhaslam I found a test that goes through different type change scenarios (https://github.com/libgit2/libgit2/blob/master/tests/checkout/typechange.c).  I'm trying to see if we could expand it to give us some more coverage of this change.

srajko commented:
I expanded the test a bit (https://github.com/srajko/libgit2/commit/4ae849f0e6a4ab4bbc6abc13f19cea28243cee36) and the test continues to pass after the change from this PR - and I took the expected values before this change.

hackhaslam commented:
Thanks @srajko. So was the change you made in srajko@c558840 was already working before this PR?

srajko commented:
Correct, that seems to have been resolved at some point, so I put in the suggested change to get tighter coverage of this PR.  I submitted a PR for your PR in case you want to include that here :-)

The tighter tests give me more confidence that this PR doesn't change libgit2 behavior with checkouts across type changes involving submodules.  @ethomson @carlosmn is there anything else you would like to see to accept this PR?

srajko commented:
I just did some manual testing, and looks like after the PR change libgit2 is letting me do a safe checkout from a dirty submodule to a tree, losing the uncommitted changes.  So maybe we need to check for a type change, and if the type is changing report a conflict if the submodule is dirty?

hackhaslam commented:
Okay, I merged your PR into my branch. The type change from submodule to tree with a dirty submodule was number 54 in the notes I made. I marked it as MAYBE SAFE, although I can't remember why now. It seems like it should practically always be a conflict. Also number 52 which core git gets wrong should probably conflict. There are at a couple of others cases (40, 41) that don't involve a dirty submodules that should also conflict.

So yeah, I think those cases need to be addressed (and tested) before this can be merged.

srajko commented:
I got the test to cover the scenarios where a submodule is dirty - PRed to your PR.

With the expected values taken before the change in this PR, the tests now fail - the actual values of conflicts are exactly one less for each checkout involving a type change from a dirty submodule. Hopefully, updating the conflict logic to take into account type changes should make the test pass again.

@hackhaslam do you have any time in the near future to update the conflict logic? If not, I'd be happy to take a stab at it.

hackhaslam commented:
I'm on vacation away from my computer for the next couple of days. I can work on it then if you don't get to it first.

srajko commented:
Pushed a potential fix to https://github.com/stinb/libgit2/pull/2

hackhaslam commented:
@srajko sorry I'm so slow to respond. I think that something like I pushed in 488ade6 might be better. It keeps the is_workdir_modified logic the same and only checks for a submodule in the modified case. I rebased it all on the tip of master and dropped my initial commit. I'm still testing some of the other cases.

BTW, I remember now why I put MAYBE SAFE on the typechange to tree case. Git only generates a conflict if the trees intersect in some way. The submodule tree basically becomes untracked, then it gives an "untracked file would be overwritten by checkout" conflict if the new tree contains any files with the same name.

srajko commented:
@hackhaslam sorry now it was me that got swept away in other things :-)

I got the CI passing on a test PR - #3734.  https://github.com/libgit2/libgit2/pull/3734/commits/8da601d383b7cc9a3113e762494bf1ec59ee6896 resolves a build error with the test code I added, and you'll probably want to cherry-pick that into this PR (or I can PR it to your branch).  When I tried that on top of this PR though, the CI still failed on something. I did get it to pass on the other PR, but that one has a slightly different diff in `src/checkout.c` (there, the bypass code you added is only applied when both the `wditem` and the `newitem` are `GIT_FILEMODE_COMMIT`)

srajko commented:
@hackhaslam Let's finish this up :-)  I think we have all the pieces, we just need to get this PR finalized and passing.  Is there anything I can do to help?

hackhaslam commented:
@srajko So is there a case that doesn't work after the changes I made? If there is then it doesn't look like it's tested.

johnhaley81 commented:
CI passes! <3 @hackhaslam and @srajko

srajko commented:
@hackhaslam perhaps the test failure I got was intermittent / unrelated, sorry for the false alarm.

I think this PR is ready.  We may want to squash https://github.com/libgit2/libgit2/pull/3583/commits/e756a84e08d4c5160917e8457ece47c483f07927 into https://github.com/libgit2/libgit2/pull/3583/commits/aee92e73ffa25f4c88d469e45fff618bcd1999cf just to clean up the history.  @carlosmn @ethomson anything else you'd like to see done here?

jdgarcia commented:
What's the status on this?

ethomson commented:
@srajko Sorry about the delay.  I really love that this is a one-line code change with so much docs and test around it.  ❤️ 

I will take a look at this later today to make sure that I understand all the documentation.  :)

hackhaslam commented:
I added new commit to address #3774. Merge has a check for modifications in the workdir that would be overwritten by the merge operation. I think that it's correct (and safe) to ignore submodules here as well.

@ethomson If this looks good I can add a changelog entry as well.

johnhaley81 commented:
I think that failure in AppVeyor was due to networking issues. @ethomson can you rerun AppVeyor please?

johnhaley81 commented:
Now that CI is 💚 is there anything else that this needs?

ethomson commented:
Sorry about the delay here; I'm manually merging this because it's been sitting for too long and it should be merged.  Apologies.  :/

In the future, we'd ask that you squash this sort of thing up into nice, standalone commits.  But I'm going to go ahead and do this one because we've been so tardy in taking care of it.

Thanks again, this is a big help.  :)
